### ISSUE TITLE
Phase 8 Rebuild: End-to-End IndicLID Runtime, Routing & API Integration

### ISSUE THREAD

**Issue by sakethalladaaa** [sakethalladaaa/malayalam-whatsapp-translator#12]
## Overview

This issue is the **fresh Phase 8 implementation tracker** for the Malayalam WhatsApp Translator.

The goal is to rebuild and verify the complete **AI4Bharat IndicLID production integration** from the beginning: reproducible runtime setup, real model inference, backend adapter/runtime loading, frozen Phase 5 routing, FastAPI integration, integration tests, CI validation, and final documentation.

This tracker is intentionally broader than the earlier Phase 8 work in #2 and #4. Existing code and test evidence may be reused **only after it is re-verified against the current `main` branch**.

Phase 9 work already merged into `main` must remain compatible; this Phase 8 rebuild must not undo or regress later pipeline work.

---

## Phase 8 Goal

For every incoming message, the backend should be able to run the real IndicLID runtime and produce a reliable language-routing decision for the existing pipeline.

```text
Input text
   ↓
Real AI4Bharat IndicLID runtime
   ↓
Language prediction
   ↓
Frozen Phase 5 candidate_v1 routing policy
   ↓
English / Malayalam-script / Roman-Malayalam / unsupported-or-mixed decision
   ↓
Existing downstream translation pipeline
```

---

## Non-negotiable Constraints

* [ ] Do not retune the frozen Phase 5 `candidate_v1` policy.
* [ ] Do not modify frozen Phase 5 evaluation datasets to improve metrics.
* [ ] Preserve Phase 6 / Phase 7 preprocessing behavior unless a regression is identified.
* [ ] Preserve compatibility with the already merged Phase 9 pipeline.
* [ ] Do not commit model binaries, caches, or virtual environments.
* [ ] IndicLID runtime failures must be explicit.
* [ ] Heavy real-model tests must remain separate from normal lightweight CI.

---

# Phase 8 Subtasks

## 8.0 — Audit Current `main`

Before changing implementation:

* [ ] Inspect `backend/app/lid.py`.
* [ ] Inspect current language routing code.
* [ ] Inspect `/translate`.
* [ ] Identify existing Phase 8 code.
* [ ] Separate production-ready code from mocks/contracts/scaffolding.
* [ ] Inspect current Phase 8 tests.
* [ ] Inspect Phase 9 dependencies on Phase 8.
* [ ] Record current backend test count.
* [ ] Record current passing/failing state.
* [ ] Verify no IndicLID model/cache files are tracked.
* [ ] Document what will be preserved, replaced, or added.

### Acceptance Criteria

Current architecture and implementation gaps are documented before new implementation begins.

---

## 8.1 — Reproducible IndicLID Runtime

* [ ] Pin the exact AI4Bharat IndicLID upstream commit.
* [ ] Pin Python version.
* [ ] Pin PyTorch.
* [ ] Pin Transformers.
* [ ] Pin FastText.
* [ ] Pin NumPy.
* [ ] Pin pandas.
* [ ] Pin other tokenizer/runtime dependencies as needed.
* [ ] Document CPU/GPU requirements.
* [ ] Make setup reproducible from a clean environment.
* [ ] Make setup safe to rerun.
* [ ] Avoid nested checkout/model paths.
* [ ] Make setup failures stop explicitly.
* [ ] Keep heavy runtime dependencies separate where appropriate.

### Acceptance Criteria

A clean environment can reproduce the IndicLID runtime using documented commands.

---

## 8.2 — Reproduce Real IndicLID Model

* [ ] Check out the pinned IndicLID revision.
* [ ] Obtain the required model artifacts.
* [ ] Initialize from a fresh runtime.
* [ ] Record Python/package/device versions.
* [ ] Verify a repeated run uses the same intended paths.
* [ ] Verify existing valid model cache behavior.
* [ ] Verify no unnecessary redownload.
* [ ] Keep model artifacts outside Git.

### Acceptance Criteria

Fresh and repeated runtime initialization both behave predictably.

---

## 8.3 — Verify Legacy Weight Loading

* [ ] Capture `load_state_dict(..., strict=False)` results.
* [ ] Inspect `missing_keys`.
* [ ] Inspect `unexpected_keys`.
* [ ] Fail on unexplained mismatches.
* [ ] Document allowed legacy differences.
* [ ] Add regression validation for accepted mismatches.

### Acceptance Criteria

Required model weights are proven to load correctly.

---

## 8.4 — Real IndicLID Inference Validation

Use the real public inference path.

* [ ] Native Malayalam input.
* [ ] Roman Malayalam input.
* [ ] English input.
* [ ] Genuine Malayalam + English mixed-script input.
* [ ] Multi-item batch.
* [ ] Assert output count equals input count.
* [ ] Assert output order equals input order.
* [ ] Assert expected labels for fixed smoke examples.
* [ ] Test repeated inference.
* [ ] Test multiple predictions with one loaded model.
* [ ] Keep direct BERT tests only for specific fallback behavior.

### Acceptance Criteria

Real `batch_predict()` behavior is validated for project-relevant inputs.

---

## 8.5 — Production Runtime Loader

* [ ] Implement production IndicLID runtime loader.
* [ ] Lazy-load the model.
* [ ] Cache/reuse one expensive model instance.
* [ ] Avoid loading the model per request.
* [ ] Use lazy imports where practical.
* [ ] Add explicit runtime-unavailable exception.
* [ ] Add explicit inference failure exception.
* [ ] Add test-only cache reset support.
* [ ] Document initialization behavior.

### Acceptance Criteria

Backend code can obtain one reusable real IndicLID engine through a stable loader.

---

## 8.6 — Backend IndicLID Adapter

* [ ] Define stable adapter input/output contract.
* [ ] Convert raw IndicLID output into project predictions.
* [ ] Preserve input text and batch order.
* [ ] Handle invalid/empty inputs.
* [ ] Convert runtime failures into project-level exceptions.
* [ ] Unit test using fake engines.
* [ ] Integration test using real engine.

### Acceptance Criteria

Application code does not depend directly on IndicLID internals.

---

## 8.7 — Frozen Phase 5 Routing Integration

* [ ] Route `mal_Mlym`.
* [ ] Route `mal_Latn`.
* [ ] Preserve Malayalam-specific Roman lexical fallback.
* [ ] Preserve English behavior.
* [ ] Define unsupported/mixed behavior.
* [ ] Run Phase 5 regression tests.
* [ ] Do not retune thresholds or markers.

### Acceptance Criteria

Real IndicLID predictions feed the existing frozen routing policy without changing its semantics.

---

## 8.8 — FastAPI Integration

Connect the real runtime to `/translate`.

* [ ] `/translate` → IndicLID adapter.
* [ ] IndicLID → frozen router.
* [ ] Preserve English passthrough.
* [ ] Preserve native Malayalam route.
* [ ] Preserve Roman Malayalam route.
* [ ] Connect downstream Phase 9 pipeline.
* [ ] Handle unsupported/mixed input according to current contract.
* [ ] Handle runtime initialization failure.
* [ ] Handle inference failure.
* [ ] Verify model is not initialized repeatedly.

### Acceptance Criteria

A real API request reaches the real IndicLID runtime and routes into the correct downstream path.

---

### Unit / Contract Tests

* [ ] Runtime loader caching.
* [ ] Runtime unavailable behavior.
* [ ] Adapter conversion.
* [ ] `mal_Mlym` routing.
* [ ] `mal_Latn` routing.
* [ ] English handling.
* [ ] Roman lexical fallback.
* [ ] Invalid/empty input.

### Real Runtime Integration Tests

* [ ] Real model initialization.
* [ ] Native Malayalam prediction.
* [ ] Roman Malayalam prediction.
* [ ] English prediction.
* [ ] Mixed-script prediction.
* [ ] Batch order/count.
* [ ] Stable repeated inference.
* [ ] Model reuse.
* [ ] Cached checkpoint initialization.
* [ ] Runtime loader → adapter integration.

### API Integration Tests

* [ ] FastAPI → IndicLID → router.
* [ ] English API path.
* [ ] Native Malayalam API path.
* [ ] Roman Malayalam API path.
* [ ] Runtime failure response.

### Acceptance Criteria

Fast tests run normally; real model tests have a documented opt-in command.

---

## 8.10 — CI and Quality Checks

Reuse current repository automation where possible.

* [ ] Phase 8 lightweight tests run on PRs.
* [ ] Ruff passes.
* [ ] CodeQL passes.
* [ ] Dependency audit passes.
* [ ] Workflow lint/config validation passes.
* [ ] Add/update IndicLID pytest marker.
* [ ] Document environment flag for heavy runtime tests.
* [ ] Ensure normal CI does not download heavy models unintentionally.
* [ ] Ensure dependency manifests are covered by maintenance tooling.

### Acceptance Criteria

Normal PR validation stays fast while real-model validation remains reproducible.

---

## 8.11 — End-to-End Phase 8 Validation

* [ ] Start from documented clean runtime.
* [ ] Initialize real IndicLID.
* [ ] Run real runtime suite.
* [ ] Run backend tests.
* [ ] Run routing regression tests.
* [ ] Run API integration tests.
* [ ] Verify Phase 9 compatibility.
* [ ] Record exact commands.
* [ ] Record test pass counts.
* [ ] Record runtime platform/device.
* [ ] Confirm model/cache artifacts are not staged.

### Acceptance Criteria

There is reproducible evidence for:

```text
Input
→ IndicLID
→ Routing
→ FastAPI
→ Existing Phase 9 pipeline
```

---

## 8.12 — Documentation and Closure

* [ ] Update README Phase 8 status.
* [ ] Document runtime installation.
* [ ] Document architecture.
* [ ] Document real-model test command.
* [ ] Document supported labels/routes.
* [ ] Document known limitations.
* [ ] Document Phase 8 → Phase 9 connection.
* [ ] Update final PR description with validation evidence.
* [ ] Link relevant PRs/commits.
* [ ] Compare final implementation with original Issue #2 requirements.
* [ ] Close this tracker only when all required acceptance criteria are complete.

---

# Definition of Done

Phase 8 is complete only when:

* [ ] IndicLID reproduces from a clean environment.
* [ ] Model-weight loading is verified.
* [ ] Real `batch_predict()` works.
* [ ] Production runtime loader exists.
* [ ] Adapter uses real runtime.
* [ ] Frozen Phase 5 routing consumes real predictions.
* [ ] FastAPI uses the real Phase 8 routing path.
* [ ] Unit tests pass.
* [ ] Real runtime integration tests pass.
* [ ] API integration tests pass.
* [ ] Lightweight CI passes.
* [ ] Heavy runtime validation is reproducible.
* [ ] Phase 9 remains compatible.
* [ ] Documentation is complete.
* [ ] No model/cache/venv artifacts are committed.

---

## Out of Scope

Phase 8 does not include:

* IndicXlit quality tuning.
* IndicTrans2 quality tuning.
* Translation-model replacement.
* UI redesign.
* Re-training IndicLID.
* Retuning the Phase 5 routing policy.

---

## Related Work

* Original Phase 8 tracker: #2
* Phase 8 contract/routing PR: #4
* Phase 9 translation pipeline: #8
* Phase 9 real IndicTrans2 runtime: #11

This issue is the new **single source of truth for rebuilding Phase 8 from baseline audit through production/API validation**.
